Infinite recursion in 2.7.0 beta on certain imports

### Describe the bug

When running dev mode on a project with certain types of imports (not sure about the exact case yet), Vite crashes with a `Maximum call stack size exceeded` error.

I don't have a ton of time, but I do plan to investigate a little bit -- if I figure it out I'll do a PR.

### Reproduction

https://github.com/davidwallacejackson/vite-fgi2rv (instructions in the repo)

I started this repro on StackBlitz, but unfortunately the behavior is different there -- instead of crashing, Vite just hangs.

### System Info

```shell
System:
    OS: macOS 11.6
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 48.86 MB / 32.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 15.14.0 - ~/.nvm/versions/node/v15.14.0/bin/node
    Yarn: 1.21.1 - /usr/local/bin/yarn
    npm: 7.7.6 - ~/.nvm/versions/node/v15.14.0/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Browsers:
    Chrome: 95.0.4638.69
    Firefox: 91.0.2
    Safari: 14.1.2
```
```


### Used Package Manager

yarn

### Logs

```shell
Error: Build failed with 1 error:
node_modules/vite/dist/node/chunks/dep-6cab8e8a.js:3977:44: error: [plugin: vite:dep-scan] Maximum call stack size exceeded
    at failureErrorWithLog (/Users/davidjackson/Downloads/vite-fgi2rv/node_modules/esbuild/lib/main.js:1493:15)
    at /Users/davidjackson/Downloads/vite-fgi2rv/node_modules/esbuild/lib/main.js:1151:28
    at runOnEndCallbacks (/Users/davidjackson/Downloads/vite-fgi2rv/node_modules/esbuild/lib/main.js:941:63)
    at buildResponseToResult (/Users/davidjackson/Downloads/vite-fgi2rv/node_modules/esbuild/lib/main.js:1149:7)
    at /Users/davidjackson/Downloads/vite-fgi2rv/node_modules/esbuild/lib/main.js:1258:14
    at /Users/davidjackson/Downloads/vite-fgi2rv/node_modules/esbuild/lib/main.js:629:9
    at handleIncomingPacket (/Users/davidjackson/Downloads/vite-fgi2rv/node_modules/esbuild/lib/main.js:726:9)
    at Socket.readFromStdout (/Users/davidjackson/Downloads/vite-fgi2rv/node_modules/esbuild/lib/main.js:596:7)
    at Socket.emit (node:events:369:20)
    at addChunk (node:internal/streams/readable:313:12)
```
```


### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/playground/worker/__tests__/worker.spec.ts
+++ b/packages/playground/worker/__tests__/worker.spec.ts
@@ -52,8 +52,8 @@ if (isBuild) {
   test('inlined code generation', async () => {
     const assetsDir = path.resolve(testDir, 'dist/assets')
     const files = fs.readdirSync(assetsDir)
-    // should have 2 worker chunk
-    expect(files.length).toBe(3)
+    // should have 3 worker chunk
+    expect(files.length).toBe(4)
     const index = files.find((f) => f.includes('index'))
     const content = fs.readFileSync(path.resolve(assetsDir, index), 'utf-8')
     const worker = files.find((f) => f.includes('my-worker'))

--- a/packages/playground/worker/__tests__/worker.spec.ts
+++ b/packages/playground/worker/__tests__/worker.spec.ts
@@ -12,6 +12,11 @@ test('normal', async () => {
   )
 })
 
+test('TS output', async () => {
+  await page.click('.ping-ts-output')
+  await untilUpdated(() => page.textContent('.pong-ts-output'), 'pong')
+})
+
 test('inlined', async () => {
   await page.click('.ping-inline')
   await untilUpdated(() => page.textContent('.pong-inline'), 'pong')

--- a/packages/playground/worker/index.html
+++ b/packages/playground/worker/index.html
@@ -7,6 +7,12 @@
 <button class="ping-inline">Ping Inline Worker</button>
 <div>Response from inline worker: <span class="pong-inline"></span></div>
 
+<button class="ping-ts-output">Ping Possible Compiled TS Worker</button>
+<div>
+  Response from worker imported from code that might be compiled TS:
+  <span class="pong-ts-output"></span>
+</div>
+
 <button class="tick-shared">Tick Shared Worker</button>
 <div>
   Tick from shared worker, it syncs between pages:
@@ -17,6 +23,7 @@
   import Worker from './my-worker?worker'
   import InlineWorker from './my-worker?worker&inline'
   import SharedWorker from './my-shared-worker?sharedworker&name=shared'
+  import TSOutputWorker from './possible-ts-output-worker?worker'
   import { mode } from './workerImport'
 
   document.querySelector('.mode-true').textContent = mode
@@ -50,6 +57,15 @@
   })
 
   sharedWorker.port.start()
+
+  const tsOutputWorker = new TSOutputWorker()
+  worker.addEventListener('message', (e) => {
+    text('.pong-ts-output', e.data.msg)
+  })
+
+  document.querySelector('.ping-ts-output').addEventListener('click', () => {
+    inlineWorker.postMessage('ping')
+  })
 
   function text(el, text) {
     document.querySelector(el).textContent = text

--- a/packages/playground/worker/possible-ts-output-worker.mjs
+++ b/packages/playground/worker/possible-ts-output-worker.mjs
@@ -1,0 +1,7 @@
+import { msg, mode } from './workerImport'
+
+self.onmessage = (e) => {
+  if (e.data === 'ping') {
+    self.postMessage({ msg, mode })
+  }
+}

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -172,7 +172,7 @@ export const isTsRequest = (url: string) => knownTsRE.test(cleanUrl(url))
 export const isPossibleTsOutput = (url: string) =>
   knownTsOutputRE.test(cleanUrl(url))
 export const getTsSrcPath = (filename: string) =>
-  filename.replace(/\.([cm])?(js)(x?)$/, '.$1ts$3')
+  filename.replace(/\.([cm])?(js)(x?)(\?|$)/, '.$1ts$3')
 
 const importQueryRE = /(\?|&)import=?(?:&|$)/
 const internalPrefixes = [

